### PR TITLE
Websocket support

### PR DESCRIPTION
**Pull request by bharel**
Full docs - should finish tomorrow. (incl. changelog, version, etc.).

### CODE REVIEW

**Comment by bharel** [comment]
Pytype keeps failing it. No support for `__future__` annotations >.<

File: yellowbox/extras/websocket.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+    def clear(self) -> None:
+        """Remove all routes."""
+        with self._lock:
+            self._routes.clear()
+            self._re_routes.clear()
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO)
```
Comment (bentheiii) [inline yellowbox/extras/websocket.py:320]: why is this necessary?

File: tests/extras/test_websocket.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+from pytest import mark, fixture, raises
+from urllib.parse import urljoin
+
+from yellowbox.extras.websocket import WebsocketService
+
+TEST_PATH = "/test"
+
+
+@fixture(scope="module")
+def global_websocket_service():
```
Comment (bentheiii) [inline tests/extras/test_websocket.py:13]: I'd underline this

Comment (bentheiii) [reply tests/extras/test_websocket.py:13]: Also why don't you make the service per-test?

File: tests/extras/test_websocket.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+
+@_parametrize_side_effect_response
+@mark.parametrize("regex", [None, "/A?test"])
+def test_websocket_route(websocket_service: WebsocketService,
+                         expected_response, side_effect, regex):
+    if regex:
+        websocket_service.route(regex=regex)(side_effect)
+    else:
+        websocket_service.route(TEST_PATH)(side_effect)
```
Comment (bentheiii) [inline tests/extras/test_websocket.py:61]: iffy. you know how to make this right

**Comment by bentheiii** [comment]
also, add `local_url` and `container_url` for convenience

File: tests/extras/test_websocket.py (hunk 3 above)
Comment (bharel) [reply tests/extras/test_websocket.py:61]: Kwargs dict? Eeww

File: yellowbox/extras/websocket.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+            raise
+
+        finally:
+            if not initialized:
+                self.close()
+
+        self._advance_generator()
+
+    def _advance_generator(
+            self, data: Union[bytearray, str, None] = None) -> None:
```
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:63]: Yeah sure

File: yellowbox/extras/websocket.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        try:
+            generator.throw(err)
+        except Exception as exc:
+            # Check if it's our exception. If it is, ignore it.
+            if isinstance(exc, ConnectionAbortedError) and exc is err:
+                return
+            logger.exception("Exception raised on connection close.")
+            raise
+        finally:
+            generator.close()  # May throw an exception. Ignore it.
```
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:98]: I do call it. This is a different exception.

File: yellowbox/extras/websocket.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+
+@no_type_check
+def _to_generator(side_effect: SIDE_EFFECT_TYPE
+                  ) -> _GEN_FUNCTION_TYPE:
```
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:103]: Auto pep8 lolz

File: yellowbox/extras/websocket.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+        Generator function, same as the one in WebsocketService.route().
+    """
+    # Side effect == normal string
+    if isinstance(side_effect, (str, bytes, bytearray, memoryview)):
+        def gen(*args: Any, **kwargs: Any) -> _GENERAOTR_TYPE:
+            return side_effect
+            yield  # On purpose.
+        return gen
+
+    @wraps(side_effect)  # type: ignore # Mypy GH-10002
```
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:119]: Tis a bug in mypy. You want me to generally assert? Cause it already throws an exception

File: yellowbox/extras/websocket.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+            if isinstance(exc, ConnectionAbortedError) and exc is err:
+                return
+            logger.exception("Exception raised on connection close.")
+            raise
+        finally:
+            generator.close()  # May throw an exception. Ignore it.
+
+
+@no_type_check
+def _to_generator(side_effect: SIDE_EFFECT_TYPE
```
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:102]: Nope. I thought about it, it's a 2 way stream. Apart from one-offs for ease of use, just make a generator. Don't think adding support for this is worth it atm

File: yellowbox/extras/websocket.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+        callback = self._routes.get(path)
+        if callback:
+            return callback
+
+        items = reversed(tuple(self._re_routes.items()))  # Thread safety
+        for pattern, callback in items:
+            if pattern.fullmatch(path):
+                return callback
+
+        return None
```
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:207]: On purpose. I'm mimicking most http server behaviors, and it's not a bug, it's intended most of the times.

File: yellowbox/extras/websocket.py (hunk 1 above)
Comment (bharel) [reply yellowbox/extras/websocket.py:320]: Made it for myself mostly as an easy server for testing and thought I'd leave it as is. Proved to be quite useful.

File: tests/extras/test_websocket.py (hunk 2 above)
Comment (bharel) [reply tests/extras/test_websocket.py:13]: For the same reason i can ask "why would I?"

File: yellowbox/extras/websocket.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bentheiii) [reply yellowbox/extras/websocket.py]: But it's an exception that you yourself throw, I don't see the added benefit of this over just calling close

File: yellowbox/extras/websocket.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bentheiii) [reply yellowbox/extras/websocket.py:126]: won't the exception only be thrown inside the runner thread?

File: yellowbox/extras/websocket.py (hunk 1 above)
Comment (bentheiii) [reply yellowbox/extras/websocket.py:327]: It should be deleted, i think

File: tests/extras/test_websocket.py (hunk 3 above)
Comment (bentheiii) [reply tests/extras/test_websocket.py:63]: what? no
```
@parameterize('uri, regex', [(TEST_PATH, None), (None, "/A?test")])
def test_(uri, regex):
  ...
  websocket_service.route(uri, regex=regex)(side_effect)
```

Comment (bharel) [reply tests/extras/test_websocket.py]: u very smurt ❤️

File: yellowbox/extras/websocket.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py:126]: Yes, it should be fine but I added an assert cause u so smauwt and I wuvs u

File: yellowbox/extras/websocket.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bharel) [reply yellowbox/extras/websocket.py]: This exception can be handled specifically + `GeneratorExit` inherits from `BaseException` so it won't call most exception handlers anyway. I be also smurt, give me some cwedit 😋

File: yellowbox/extras/websocket.py (hunk 1 above)
Comment (bharel) [reply yellowbox/extras/websocket.py:327]: why? 'tis echo server in websocket.
can be good for testing and vewy comfowtable. you no think?

File: yellowbox/extras/websocket.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bentheiii) [reply yellowbox/extras/websocket.py:128]: It won't be fine, exceptions thrown in the runner thread are swallowed and return 500, sad!

File: yellowbox/extras/websocket.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+        def gen(*args: Any, **kwargs: Any) -> _GENERAOTR_TYPE:
+            return side_effect
+            yield  # On purpose.
+        return gen
+    
+    # Side effect == list of strings
+    if isinstance(side_effect, Iterable):
+        def gen(*args, **kwargs) -> _GENERAOTR_TYPE:
+            for item in side_effect:
+                yield item
```
Comment (bentheiii) [inline yellowbox/extras/websocket.py:123]: be nice to fellow generators
```suggestion
            return (yield from side_effect)
```

File: yellowbox/extras/websocket.py (hunk 1 above)
Comment (bentheiii) [reply yellowbox/extras/websocket.py:329]: I don't think we want to encourage people running our modules as mains

**Review by bentheiii — APPROVED** [review]

File: yellowbox/extras/websocket.py (hunk 10 above)
Comment (bharel) [reply yellowbox/extras/websocket.py:144]: I would, if it wouldn't cause bugs
